### DISCUSSION TITLE
JsonLd does not work in NextJs 13 App Router

### DISCUSSION THREAD

**Discussion by KMadhanSPM** [vercel/next.js#56659] [Help]
### Summary

    Need to render schema in Head tag. Unable to render it. Can you please explain?

**Comment by icyJoseph** [comment]
What have you tried so far?

Their recommendation is to place it in a script tag wherever you need it:

- https://nextjs.org/docs/app/building-your-application/optimizing/metadata#json-ld

I tried it with their code, placed it within a `head` tag at the root layout, and it worked, but I think that's not ideal. And it might not be possible for you. Why would it need to be in the head tag?

**Reply by KMadhanSPM** [reply]
I tried the above one, what you shared. But it is rendering in body tag.

I need to render in head tag because JSON-LD is just one form of structured data, and its placement in the <head> is essential for search engines to effectively understand and process the data.

**Reply by icyJoseph** [reply]
Mmm according to, https://jsonld.com/why-json-ld/#:~:text=It%20is%20generally%20safe%20to,verification's%20through%20Google's%20testing%20tool.:

> Where do I insert this code?
> 
> It is generally safe to insert JSON-LD code within the `<head>` of your website, as such this is what we recommend. JSON-LD can also be inserted in the `<body>` of a website, or anywhere else for that matter.
> 
> We’ve confirmed this by running several structured data verification’s through Google’s testing tool.

And on my closing line, I said that I tried wrapping a head tag on the above example, and it worked. Have you tried that? Then again, the JSON-LD can also be inserted in the body as their own documents indicate.

**Reply by icyJoseph** [reply]
Last but not least, although I think it is not very safe to inject these directly into the head yourself, you can do either of these:

- https://stackblitz.com/edit/nextjs-ch23gk?file=app%2Flayout.tsx (direct head injection)
- https://stackblitz.com/edit/nextjs-tf9jjc?file=app%2Flayout.tsx (next/script with beforeInteractive strategy)

**Reply by KMadhanSPM** [reply]
@icyJoseph We are having dynamic pages using slug. In that we are fetching schema based on route URL. 

If we are rendering schema in layout.jsx, we are not able to get the dynamic data. Can you give the solution for this?

**Reply by mmspring13** [reply]
use [useServerInsertedHTML](https://nextjs.org/docs/app/building-your-application/styling/css-in-js#configuring-css-in-js-in-app) for that

**Reply by bdlowery** [reply]
> Last but not least, although I think it is not very safe to inject these directly into the head yourself, you can do either of these:
> 
> * https://stackblitz.com/edit/nextjs-ch23gk?file=app%2Flayout.tsx (direct head injection)
> * https://stackblitz.com/edit/nextjs-tf9jjc?file=app%2Flayout.tsx (next/script with beforeInteractive strategy)

There's not a single instance in the docs that has an example using `<head>` in app router. You probably shouldn't do it.
The second one causes hydration errors.

**Reply by 4anghyeon** [reply]
> > Last but not least, although I think it is not very safe to inject these directly into the head yourself, you can do either of these:
> > 
> > * https://stackblitz.com/edit/nextjs-ch23gk?file=app%2Flayout.tsx (direct head injection)
> > * https://stackblitz.com/edit/nextjs-tf9jjc?file=app%2Flayout.tsx (next/script with beforeInteractive strategy)
> 
> There's not a single instance in the docs that has an example using `<head>` in app router. You probably shouldn't do it. The second one causes hydration errors.

using ⁠next/script with the ⁠beforeInteractive strategy, is the correct approach.
The hydration error occurs because the `Script` tag is placed directly below the `⁠html` tag. It should be moved inside the ⁠`body` tag.


[According to the Next.js documentation:](https://nextjs.org/docs/app/api-reference/components/script#beforeinteractive)
> Good to know: Scripts with ⁠beforeInteractive will always be injected inside the head of the HTML document regardless of where it's placed in the component.

**Comment by Uriyo** [comment]
I have also the same use case in which i am trying to add json ld for my different blog pages but the thing that is given in the docs does not works for me too i am on next:14   [](https://nextjs.org/docs/app/guides/json-ld)

**Comment by Uriyo** [comment]
Okay I have founhd solution for me 
Try using next/script tag provided by the next js this worked for me with ssg


```
 <Script
        id="article-schema"
        type="application/ld+json"
        strategy="beforeInteractive"
        dangerouslySetInnerHTML={{
          __html: JSON.stringify(articleSchema).replace(/</g, '\\u003c'),
        }}
      />
```

```

const articleSchema = {
      "@context": "https://schema.org",
      "@type": "Article",
      "headline": blog.title,
      "image": blog.image_url,  
      "author": {
        "@type": "Person",
        "name": blog.author_name
      },  
      "publisher": {
        "@type": "Organization",
        "name": "Prepzo",
        "logo": {
          "@type": "ImageObject",
          "url": "https://www.prepzo.ai/logo.png"
        }
      },
      "datePublished": blog.publish_date
  };
```

**Comment by jonataspinto** [comment]
"next": "^14.2.26" with app router.

```javascript
// /public/next/scripts/ldJSON.js
(function (){
    if (typeof window !== 'undefined') {
        window.ldJSON = (data) => {
        var script = document.getElementById('schema');
        if (script) {
            script.innerHTML = data;
            return;
        }
        var node = document.getElementsByTagName('script')[0];
        script = document.createElement('script');
        script.key = 'structured-data';
        script.type = 'application/ld+json';
        script.id = 'schema';
        script.innerHTML = data;
        node.parentNode.insertBefore(script, node);
    }
    }
})()

```


```jsx
'use client';

import Script from 'next/script';

export function JsonLdScript({ schema }: { schema: Record<string, unknown> }) {
  return (
    <Script
      src="/next/scripts/ldJSON.js"
      onReady={() => {
        if (window?.ldJSON) {
          window?.ldJSON(JSON.stringify(schema));
        }
      }}
    />
  );
}
```
